src/coreComponents/constitutive/fluid/PVTFunctions/SpanWagnerCO2DensityFunction.hpp

```diff
@@ -13,61 +13,159 @@
+ * @file SpanWagnerCO2Density.hpp
  */
 
-#ifndef GEOSX_CONSTITUTIVE_FLUID_PVTFUNCTIONS_SPANWAGNERCO2DENSITYFUNCTION_HPP_
-#define GEOSX_CONSTITUTIVE_FLUID_PVTFUNCTIONS_SPANWAGNERCO2DENSITYFUNCTION_HPP_
+#ifndef GEOSX_CONSTITUTIVE_FLUID_PVTFUNCTIONS_SPANWAGNERCO2DENSITY_HPP_
+#define GEOSX_CONSTITUTIVE_FLUID_PVTFUNCTIONS_SPANWAGNERCO2DENSITY_HPP_
 
 #include "PVTFunctionBase.hpp"
 
+#include "constitutive/fluid/PVTFunctions/PVTFunctionHelpers.hpp"
+#include "managers/Functions/TableFunction.hpp"
+
 namespace geosx
 {
 
+namespace constitutive
+{
+
 namespace PVTProps
 {
 
-class SpanWagnerCO2DensityFunction : public PVTFunction
+class SpanWagnerCO2DensityUpdate final : public PVTFunctionBaseUpdate
 {
 public:
 
+  SpanWagnerCO2DensityUpdate( arrayView1d< string const > const & componentNames,
+                              arrayView1d< real64 const > const & componentMolarWeight,
+                              TableFunction * CO2DensityTable,
+                              localIndex const CO2Index )
+    : PVTFunctionBaseUpdate( componentNames,
+                             componentMolarWeight ),
+    m_CO2DensityTable( CO2DensityTable->createKernelWrapper() ),
+    m_CO2Index( CO2Index )
+  {}
 
-  SpanWagnerCO2DensityFunction( string_array const & inputPara,
-                                string_array const & componentNames,
-                                real64_array const & componentMolarWeight );
+  /// Default copy constructor
+  SpanWagnerCO2DensityUpdate( SpanWagnerCO2DensityUpdate const & ) = default;
 
-  ~SpanWagnerCO2DensityFunction() override {}
+  /// Default move constructor
+  SpanWagnerCO2DensityUpdate( SpanWagnerCO2DensityUpdate && ) = default;
 
+  /// Deleted copy assignment operator
+  SpanWagnerCO2DensityUpdate & operator=( SpanWagnerCO2DensityUpdate const & ) = delete;
 
-  static constexpr auto m_catalogName = "SpanWagnerCO2Density";
-  static string catalogName()                    { return m_catalogName; }
-  virtual string getCatalogName() const override final { return catalogName(); }
+  /// Deleted move assignment operator
+  SpanWagnerCO2DensityUpdate & operator=( SpanWagnerCO2DensityUpdate && ) = delete;
+
+  GEOSX_HOST_DEVICE
+  GEOSX_FORCE_INLINE
+  virtual void compute( real64 const & pressure,
+                        real64 const & temperature,
+                        arraySlice1d< real64 const > const & phaseComposition,
+                        real64 & value,
+                        real64 & dValue_dPressure,
+                        real64 & dValue_dTemperature,
+                        arraySlice1d< real64 > const & dValue_dPhaseComposition,
+                        bool useMass = 0 ) const override;
+
+protected:
+
+  /// Table with brine density tabulated as a function (P,T,sal)
+  TableFunction::KernelWrapper const m_CO2DensityTable;
+
+  /// Index of the CO2 component
+  localIndex const m_CO2Index;
+
+};
+
+class SpanWagnerCO2Density : public PVTFunctionBase
```

Comment:
The file name still has "Function" in it and doesn't match the class name, could you rename the file?
(I agree with removing "Function" from class names, just to be clear)

Reply:
Done, I renamed this file, and all the others that you mentioned in your review.